Star Rating/Star Rating/C/CustomControl.swift

```diff
@@ -0,0 +1,172 @@
+            star.frame.size = CGSize(width: componentDimension, height: componentDimension)
+            star.text = "★"
+            star.font = .boldSystemFont(ofSize: 35)
+            star.textAlignment = .center
+            
+            stackView.addArrangedSubview(star)
+            stackView.translatesAutoresizingMaskIntoConstraints = false
+            
+            //Set Star Color
+            switch star.tag {
+            case 1:
+                star.textColor = componentActiveColor
+            case 2:
+                star.textColor = componentInactiveColor
+            case 3:
+                star.textColor = componentInactiveColor
+            case 4:
+                star.textColor = componentInactiveColor
+            case 5:
+                star.textColor = componentInactiveColor
+            default:
+                star.textColor = componentInactiveColor
+            }
+            
+            starArray.append(star)
+        }
+    }
+}
+
+//MARK: Touch Controls
+extension CustomControl {
+    
+    func updateValue(at touch: UITouch) {
+        
+        let touchPoint = touch.location(in: self)
+        
+        for starLabel in starArray {
+            if starLabel.frame.contains(touchPoint) {
+                value = starLabel.tag
+                sendActions(for: .valueChanged)
+            
+                for starLabel in starArray {
+                    if starLabel.tag <= value {
+                        starLabel.textColor = componentActiveColor
+                        starLabel.performFlare()
+                    } else {
+                        starLabel.textColor = componentInactiveColor
+                    }
+                }
+            }
+        }
+        
+    }
+    
+    override func beginTracking(_ touch: UITouch, with event: UIEvent?) -> Bool {
+        
+        updateValue(at: touch)
+        
+        return true
+    }
+    
+    override func continueTracking(_ touch: UITouch, with event: UIEvent?) -> Bool {
+        
+        let touchPoint = touch.location(in: self)
+        
+        if bounds.contains(touchPoint) {
+            updateValue(at: touch)
+            sendActions(for: [.touchDragInside, .valueChanged])
+        } else {
+            sendActions(for: .touchDragOutside)
+        }
+        
+        return true
+    }
+    
+    override func endTracking(_ touch: UITouch?, with event: UIEvent?) {
+        
+        defer { super.endTracking(touch, with: event) }
+        
+        guard let touchPoint = touch?.location(in: self) else { return }
+        
+        if bounds.contains(touchPoint) {
+            sendActions(for: [.touchUpInside, .valueChanged])
+        } else {
+            sendActions(for: .touchUpOutside)
+        }
+        
+    }
+    
+    override func cancelTracking(with event: UIEvent?) {
+        
+        sendActions(for: [.touchCancel])
+    }
+}
+
+extension UIView {
+    
+    func performFlare() {
+        
+        func flare() {
```

Comment:
this is really interesting! I forgot you can actually nest functions! However, for readability, I'd suggest making these into closures instead:
```
let flare = {
    transform = CGAffineTransform(scaleX: 1.6, y: 1.6)
}
```

functionally it'll be the same, but the code is a bit more clear in terms of readability.